sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JacksonParser.scala

```diff
@@ -35,184 +34,289 @@ import org.apache.spark.util.Utils
+        case VALUE_STRING =>
+          // This one will lose microseconds parts.
+          // See https://issues.apache.org/jira/browse/SPARK-10681.
+          DateTimeUtils.stringToTime(parser.getText).getTime * 1000L
 
-      case (VALUE_NUMBER_INT, LongType) =>
-        parser.getLongValue
+        case VALUE_NUMBER_INT =>
+          parser.getLongValue * 1000000L
+      }
 
-      case (VALUE_TRUE, BooleanType) =>
-        true
+    case DateType =>
+      (parser: JsonParser) => parseJsonToken(parser, dataType) {
+        case VALUE_STRING =>
+          val stringValue = parser.getText
+          if (stringValue.contains("-")) {
+            // The format of this string will probably be "yyyy-mm-dd".
+            DateTimeUtils.millisToDays(DateTimeUtils.stringToTime(parser.getText).getTime)
+          } else {
+            // In Spark 1.5.0, we store the data as number of days since epoch in string.
+            // So, we just convert it to Int.
+            stringValue.toInt
+          }
+      }
+
+    case BinaryType =>
+      (parser: JsonParser) => parseJsonToken(parser, dataType) {
+        case VALUE_STRING => parser.getBinaryValue
+      }
 
-      case (VALUE_FALSE, BooleanType) =>
-        false
+    case dt: DecimalType =>
+      (parser: JsonParser) => parseJsonToken(parser, dataType) {
+        case (VALUE_NUMBER_INT | VALUE_NUMBER_FLOAT) =>
+          Decimal(parser.getDecimalValue, dt.precision, dt.scale)
+      }
 
-      case (START_OBJECT, st: StructType) =>
-        convertObject(factory, parser, st)
+    case st: StructType =>
+      val fieldConverters = st.map(_.dataType).map(makeConverter)
+      (parser: JsonParser) => parseJsonToken(parser, dataType) {
+        case START_OBJECT => convertObject(parser, st, fieldConverters)
+      }
 
-      case (START_ARRAY, ArrayType(st, _)) =>
-        convertArray(factory, parser, st)
+    case at: ArrayType =>
+      val elementConverter = makeConverter(at.elementType)
+      (parser: JsonParser) => parseJsonToken(parser, dataType) {
+        case START_ARRAY => convertArray(parser, elementConverter)
+      }
 
-      case (START_OBJECT, MapType(StringType, kt, _)) =>
-        convertMap(factory, parser, kt)
+    case mt: MapType =>
+      val valueConverter = makeConverter(mt.valueType)
+      (parser: JsonParser) => parseJsonToken(parser, dataType) {
+        case START_OBJECT => convertMap(parser, valueConverter)
+      }
 
-      case (_, udt: UserDefinedType[_]) =>
-        convertField(factory, parser, udt.sqlType)
+    case udt: UserDefinedType[_] =>
+      makeConverter(udt.sqlType)
 
-      case (token, dataType) =>
-        // We cannot parse this token based on the given data type. So, we throw a
-        // SparkSQLJsonProcessingException and this exception will be caught by
-        // parseJson method.
-        throw new SparkSQLJsonProcessingException(
-          s"Failed to parse a value for data type $dataType (current token: $token).")
+    case _ =>
```

Comment:
Or please let me correct this as below:

```
parseJsonToken(parser, dataType)(PartialFunction.empty)
```

So that they can be caught in `parseJsonToken` together with documentation.